### DISCUSSION TITLE
Is there a function to detect k-complexes for N2 detection?

### DISCUSSION THREAD

**Discussion by PhD-GOAT** [raphaelvallat/yasa#152] [Q&A] [answered]
asking because there is a spindle detection implementation to detect N2 sleep, but not a k-complex! I wonder if there is a development plan.

**Comment by raphaelvallat** [comment, marked as answer]
Hi @PhD-GOAT,

You should be able to detect k-complexes with the [yasa.sw_detect](https://raphaelvallat.com/yasa/build/html/generated/yasa.sw_detect.html) function: you just need to make sure that you limit the detection to N2 sleep.

**Reply by PhD-GOAT** [reply]
Oh I see, so is there a way to graph mask spindle and kcomplex at the same time on N2? When I tried it on real clinical data, it seems to detect kcomplex well but not spindle well.

**Reply by raphaelvallat** [reply]
No, you will need to apply the yasa.sw_detect and yasa.spindles_detect function separately. Have you tried adjusting the spindles parameters to get better sensitivity for sleep spindles detection?

**Reply by raphaelvallat** [reply]
I think you may have to find better parameters for the spindles detection. See https://raphaelvallat.com/yasa/build/html/faq.html#best_params

**Reply by PhD-GOAT** [reply]
Okay, I understand your intent. As an additional question, is there any way to count spindles or k-complexes for every 30 second epoch? When I do the algorithm every 30 seconds, I get AttributeError: 'NoneType' object has no attribute 'summary'. I have a problem because I need to see the entire count with the sw.summary() function for the entire data.

**Reply by raphaelvallat** [reply]
Applying the spindles detection on only 30-seconds of data at a time is not a good idea. Instead, you should apply on the entire night, and then use the [summary](https://raphaelvallat.com/yasa/build/html/generated/yasa.SpindlesResults.html#yasa.SpindlesResults.summary) method, which will give you the start times of all the detected spindles. From there, you should be able to use a pandas.Groupby operation to count the number of spindles in 30-sec epochs (e.g. [pandas.Groupby.count](https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.DataFrameGroupBy.count.html))

**Reply by PhD-GOAT** [reply]
Good. But what if some PSGs don't detect the K-complex and give an error? Of course, we're running the whole time.

**Reply by raphaelvallat** [reply]
Sorry I'm not sure I understand the question. Have you seen errors when running the SW detection?

**Reply by PhD-GOAT** [reply]
Yes. I mean, if I apply a function that detects K complex for the whole time, it fails to detect in a few psg.

**Reply by PhD-GOAT** [reply]
And there seems to be a fatal error in detecting the kcomplex with the sw function. The kcomplex should have a positive peak followed by a negative peak, but the sw function starts with a negative peak. So I tried amp setting -, but it didn't work.

**Reply by raphaelvallat** [reply]
YASA slow-wave detection expects a large negative peak followed by a small positive peak, not the other way around. In that case you might be better off not using YASA and trying to find another detection method for K-complexes. Manual scoring might be the most accurate method.
